docs/proposals/ns-scoped-catalog.md

```diff
@@ -0,0 +1,256 @@
+
+- Make it possible to use a service and plan from namespace X to provision a `ServiceInstance` in namespace Y
+- Allow creating a `ServiceBinding` in namespace X to a `ServiceInstance` in namespace Y
+- Expose a service and plan in namespace X to the cluster scope
+- Expose a service and plan in namespace X to namespace Y
+- Create a policy that adds certain services and plans to multiple namespaces
+- Creating a virtual resource that allows users to see all services or plans available to them across the cluster and namespace scopes
+
+We should take care to achieve the goals of this proposal without preventing further progress on other issues that are out of scope.
+
+## Analysis
+
+### Why namespaces?
+
+Unfortunately, it is not possible in Kubernetes to create an ACL (access control list) filtering scheme that shows users only the resources in a collection that they are allowed to see. The fundamental gaps here are:
+
+1. Not all authorizers are able to provide a list of subjects with access to a resource
+2. An external authorizer may have its state changed at any time out of band to kubernetes, making it impossible to do implement a correct `LIST` or `WATCH` operation from a certain resource version
+
+Since ACL-filtering the cluster-scoped list of services and plans is not a realistic option, we must find another method of controlling read/write access to resources. In Kubernetes, namespaces are the defacto way of performing this type of access control.
+
+Adding namespaced resources for service brokers (`ClusterServiceBroker`), services (`ClusterServiceClass`) and plans (`ClusterServicePlans`) allows us to take advantage of the existing namespace concept to perform access control.
+
+### Filtering services and plans from a broker
+
+Adding namespaced resources for brokers, services and plans is necessary but not sufficient to control access to services and plans. A single broker may offer a mix of services that all users should be able to access and services that should only be usable by some users.
+
+In order to prevent a broker that offers a mix of unprivileged and privileged services to the cluster-scoped catalog, there must be a way to filter the services and plans exposed by a broker. This can be accomplished through the use of white/black lists that control which services and plans in a broker's catalog have Service Catalog resources created for them. For example:
+
+- A cluster administrator should be able to prevent privileged services from appearing in cluster-scoped catalog
+- A cluster administrator should be able to add certain privileged services to a namespace
+
+## Design
+
+In this proposal we'll focus on adding the namespaced `ServiceBroker`, `ServiceClass`, and `ServicePlan` resources. For details on filtering which services and plans in a broker's catalog have k8s resources created for them, see https://github.com/kubernetes-incubator/Service Catalog/pull/1773.
```

Comment:
@pmorie I have a general question about namespaced brokers. 
Does it mean I can register same service-broker many times? For example Azure Broker as ClusterServiceBroker, and then as ServiceBroker twice for particular namespaces.

Reply:
Also have the same question. Seems it should be possible to have the same broker as a CSB and a SB but would be good to know for sure

Reply:
Yes, you can absolutely do that. I'd be happy to make that explicit if people care about the design proposal beyond this PR.

Reply:
Would this perhaps be handled by the catalog restrictions filter on the broker? IE if using the same broker for a service broker and cluster broker then set a filter https://github.com/kubernetes-incubator/service-catalog/pull/1773/files#diff-316f664334f2a9051f4897bee901c5f5R135 to filter out what you want to expose as a ServiceClass and what you want to expose as a ClusterServicePlan. An example of doing this would be very useful I think (If this is the way it would be done)

Reply:
@zihantang-rh, IMO when getting list of service classes you will use different API to distinguish same class with same uuid by SB or CSB API. During provisioning you always provide clusterServiceClassExternalName or serviceClassExternalName.